File: services/api/app/interface/schemas/lang_id.py
from __future__ import annotations

import datetime as dt

from pydantic import BaseModel, ConfigDict, Field


class LanguageLabelIn(BaseModel):
    confirmed_language: str | None = None
    corpus_split: str | None = None


class LabelProgressOut(BaseModel):
    total: int
    labeled: int
    unlabeled: int
    train_count: int
    test_count: int


class AutoSplitResultOut(BaseModel):
    train_assigned: int
    test_assigned: int


class LangIdProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    method: str
    language: str | None
    source_document_count: int
    source_char_count: int
    built_at: dt.datetime


class BuildLexicalProfileRequest(BaseModel):
    language: str


class IdentificationOutcomeOut(BaseModel):
    method: str
    predicted_language: str
    distances: dict[str, float]
    elapsed_ms: float


class IdentifyDocumentRequest(BaseModel):
    document_id: int
    methods: list[str] | None = None


class IdentifyUrlRequest(BaseModel):
    url: str
    methods: list[str] | None = None


class IdentifyTextRequest(BaseModel):
    text: str
    is_html: bool = False
    methods: list[str] | None = None


class IdentifyResponseOut(BaseModel):
    results: list[IdentificationOutcomeOut]


class LangIdTrainingJobOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: str
    epochs_total: int
    epochs_completed: int
    current_loss: float | None
    current_train_accuracy: float | None
    error_message: str | None
    created_at: dt.datetime
    started_at: dt.datetime | None
    finished_at: dt.datetime | None


class LangIdRunCreate(BaseModel):
    collection_id: int
    methods: list[str] = Field(min_length=1)


class LangIdRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    collection_id: int
    method: str
    status: str
    documents_total: int
    documents_processed: int
    error_message: str | None
    created_at: dt.datetime
    started_at: dt.datetime | None
    finished_at: dt.datetime | None


class LangIdResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: int
    document_id: int
    predicted_language: str
    distances: dict[str, float]
    elapsed_ms: float
    is_correct: bool | None


class LangIdRunSummaryOut(BaseModel):
    run_id: int
    method: str
    documents_evaluated: int
    accuracy: float
    precision: float
    recall: float
    f1: float
    mean_elapsed_ms: float
    confusion: dict[str, dict[str, int]]


class LangIdCompareResponseOut(BaseModel):
    summaries: list[LangIdRunSummaryOut]

File: frontend/src/api/types/langId.ts
export type LangIdMethod = "frequent_words" | "alphabetic" | "neural";

export const LANG_ID_METHODS: readonly LangIdMethod[] = ["frequent_words", "alphabetic", "neural"];

export const LANG_ID_METHOD_LABELS: Record<LangIdMethod, string> = {
  frequent_words: "Частотные слова",
  alphabetic: "Алфавитный",
  neural: "Нейросетевой",
};

export const LANG_ID_LANGUAGES: readonly { code: string; label: string }[] = [
  { code: "fr", label: "Французский" },
  { code: "en", label: "Английский" },
];

export interface LabelProgress {
  total: number;
  labeled: number;
  unlabeled: number;
  train_count: number;
  test_count: number;
}

export interface AutoSplitResult {
  train_assigned: number;
  test_assigned: number;
}

export interface LangIdProfile {
  id: number;
  method: LangIdMethod;
  language: string | null;
  source_document_count: number;
  source_char_count: number;
  built_at: string;
}

export interface IdentificationOutcome {
  method: LangIdMethod;
  predicted_language: string;
  distances: Record<string, number>;
  elapsed_ms: number;
}

export interface IdentifyResponse {
  results: IdentificationOutcome[];
}

export type LangIdJobStatus = "pending" | "running" | "completed" | "failed";

export const TERMINAL_LANG_ID_JOB_STATUSES: readonly LangIdJobStatus[] = ["completed", "failed"];

export interface LangIdTrainingJob {
  id: number;
  status: LangIdJobStatus;
  epochs_total: number;
  epochs_completed: number;
  current_loss: number | null;
  current_train_accuracy: number | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export interface LangIdRun {
  id: number;
  collection_id: number;
  method: LangIdMethod;
  status: LangIdJobStatus;
  documents_total: number;
  documents_processed: number;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export interface LangIdResult {
  id: number;
  run_id: number;
  document_id: number;
  predicted_language: string;
  distances: Record<string, number>;
  elapsed_ms: number;
  is_correct: boolean | null;
}

export interface LangIdRunSummary {
  run_id: number;
  method: LangIdMethod;
  documents_evaluated: number;
  accuracy: number;
  precision: number;
  recall: number;
  f1: number;
  mean_elapsed_ms: number;
  confusion: Record<string, Record<string, number>>;
}

export interface LangIdCompareResponse {
  summaries: LangIdRunSummary[];
}
